### PR TITLE
Adds support to Distributed closest point query for empty ranks 

### PR DESCRIPTION
**Pull request by kennyweiss**
# Summary

- This PR is a feature
- It allows quest's `DistributedClosestPoint` query to be run on meshes where some ranks have empty "object meshes" and/or "query meshes". 
- This matches how users are planning to run the code, where the "object meshes" might be features that exist on only a subset of the ranks, and similarly, where the query points might also only exist on a subset of the ranks.
- I tried to minimize changes to the `DistributedClosestPoint` class in anticipation of merging #879 (I'm not sure which one will get merged first)

The following picture was run on a 36 rank problem where ~25% of the ranks did not have any object points (the points on the circle) and, independently, ~25% of the ranks did not send any data to the distributed closest point query. This run also used a cutoff distance of .5 units
dcp_36ranks_20percentoff_ 5cutoff

The command line for this run was:
> \>srun -n36 ./bin/data_collection_util -p1 --min -1.5 -1.5 --max 1.5 1.5 --res 1000 1000
> \>srun -n36 ./examples/quest_distributed_distance_query_ex -m box_2d.root -n 1000 --empty-rank-probability=.25 -d .5

### CODE REVIEW

**Review by rhornung67 — APPROVED** [review]

File: src/axom/quest/examples/quest_distributed_distance_query_example.cpp [hunk 1]
```diff
@@ -459,13 +496,79 @@ class ObjectMeshWrapper
...
+      SLIC_DEBUG_IF(
+        m_verbose,
+        axom::fmt::format("After all gather: [{}]", axom::fmt::join(arr, ",")));
+
+      axom::Array<int> sums(numRanks + 1, numRanks + 1);
+      sums[0] = 0;
+      for(int i = 1; i <= numRanks; ++i)
+      {
+        sums[i] = sums[i - 1] + arr[i - 1];
+      }
```
Comment (kennyweiss) [inline src/axom/quest/examples/quest_distributed_distance_query_example.cpp:522]: Note: This is a scan, but the distributed closest point algorithm doesn't require RAJA, so I couldn't use the RAJA version.

Comment (rhornung67) [reply src/axom/quest/examples/quest_distributed_distance_query_example.cpp:522]: RAJA scans should be used for OpenMP, CUDA, and HIP, correct?

Comment (kennyweiss) [reply src/axom/quest/examples/quest_distributed_distance_query_example.cpp:522]: True, but this one's just on the host -- it's scanning a single scalar per rank -- so not really worth the transfer cost, or requiring a strict dependency on RAJA for the query.

**Review by agcapps — APPROVED** [review]
Thank you, @kennyweiss .

File: src/axom/quest/DistributedClosestPoint.hpp [hunk 2]
```diff
@@ -442,8 +442,12 @@ class DistributedClosestPointImpl
       conduit::relay::io::save(n, fname, protocol);
     };
 
+    const bool has_query_points = mesh_node.has_child("coordsets");
+
     // create conduit node containing data that has to xfer between ranks
+    // xfer_node will be empty if there are no query points on this rank
```
Comment (gunney1) [inline src/axom/quest/DistributedClosestPoint.hpp:448]: What if "coordsets" is present but has size zero?  Will it be a problem that `coords` is empty?  I think yes, but need to confirm.

File: src/axom/quest/examples/CMakeLists.txt [hunk 3]
```diff
@@ -52,14 +52,18 @@ if(ENABLE_MPI AND MFEM_FOUND AND MFEM_USE_MPI
     if(AXOM_ENABLE_TESTS AND AXOM_DATA_DIR)
         set(_nranks 3)
 
+        # Bugfix: Serialize ordering of DCP tests as workaround for MPI issues 
+        # encountered when these tests ran in parallel
+        set(_dcp_previous_test)
```
Comment (gunney1) [inline src/axom/quest/examples/CMakeLists.txt:57]: What exactly was the problem?  `data_collection_util` instances writing to the same filenames?

File: src/axom/quest/examples/quest_distributed_distance_query_example.cpp [hunk 4]
```diff
@@ -191,7 +207,7 @@ struct BlueprintParticleMesh
   /// Returns true if points have been added to the particle mesh
   bool hasPoints() const
   {
-    return m_coordsGroup != nullptr && m_coordsGroup->hasGroup("values");
+    return m_coordsGroup != nullptr && m_coordsGroup->hasView("values/x");
```
Comment (gunney1) [inline src/axom/quest/examples/quest_distributed_distance_query_example.cpp:210]: Would the `hasGroup` and `hasView` calls as coded ever return different values?  Just curious about why the change.  I was assuming that correct blueprint data means that if "values" exist, so does "values/x".

File: src/axom/quest/examples/quest_distributed_distance_query_example.cpp [hunk 5]
```diff
@@ -238,29 +245,46 @@ struct BlueprintParticleMesh
 
     const int SZ = pts.size();
 
+    m_dimension = NDIMS;
+
+    // lamda to create a strided view into the buffer
+    // uses workaround for empty meshes since apply() requires size > 0
+    auto createAndApplyView = [=](sidre::Group* grp,
```
Comment (gunney1) [inline src/axom/quest/examples/quest_distributed_distance_query_example.cpp:252]: This is cool.

**Review by gunney1 — APPROVED** [review]
Thanks @kennyweiss  I had some questions, but only for my own knowledge.

File: src/axom/quest/examples/CMakeLists.txt (hunk 3 above)
Comment (kennyweiss) [reply src/axom/quest/examples/CMakeLists.txt:57]: I think the problem is that they're all using the default communicator with custom tags and messages between instances of the distributed closest point example were getting crossed.

I opened #889 to allow users to register a custom communicator, and _\<speculation\>_ I think once we use `MPI_Comm_split` this problem will go away _\</speculation\>_

File: src/axom/quest/examples/quest_distributed_distance_query_example.cpp (hunk 4 above)
Comment (kennyweiss) [reply src/axom/quest/examples/quest_distributed_distance_query_example.cpp:210]: It was mostly because I was actually grabbing `"values/x"` a few lines below, so it's a bit more of a safeguard.

File: src/axom/primal/operators/squared_distance.hpp [hunk 6]
```diff
@@ -69,32 +71,30 @@ AXOM_HOST_DEVICE inline double squared_distance(const Point<T, NDIMS>& A,
...
+ * \return the squared distance from P to the closest point on box \a B
+ * or std::numeric_limits<T>::max() if \a B is invalid. 
  */
 template <typename T, int NDIMS>
 AXOM_HOST_DEVICE inline double squared_distance(const Point<T, NDIMS>& P,
                                                 const BoundingBox<T, NDIMS>& B)
 {
+  using axom::utilities::clampVal;
+
+  if(!B.isValid())
```
Comment (kennyweiss) [inline src/axom/primal/operators/squared_distance.hpp:83]: Bugfix -- we now account for invalid bounding boxes in `squared_distance(point, bbox)`

Comment (kennyweiss) [reply src/axom/primal/operators/squared_distance.hpp:83]: Value judgement -- for invalid bounding boxes, we now return `std::numeric_limits<T>::max()` instead of NaN.
This seemed less likely to get in users way, since we don't need to always check for `isNaN`.

Thoughts?

File: src/axom/primal/operators/squared_distance.hpp [hunk 7]
```diff
@@ -131,7 +131,7 @@ AXOM_HOST_DEVICE inline double squared_distance(const BoundingBox<T, NDIMS>& A,
     return v.squared_norm();
   }
 
-  return std::numeric_limits<T>::quiet_NaN();
+  return std::numeric_limits<T>::max();
```
Comment (kennyweiss) [inline src/axom/primal/operators/squared_distance.hpp:134]: Value judgement -- for invalid bounding boxes, we now return `std::numeric_limits<T>::max()` instead of NaN in `squared_distance(bbox,bbox)`.
This seemed less likely to get in users way, since we don't need to always check for `isNaN()`.

Thoughts?

File: src/axom/quest/DistributedClosestPoint.hpp [hunk 8]
```diff
@@ -141,13 +146,17 @@ template <int NDIMS>
 void get_bounding_box_from_conduit_node(primal::BoundingBox<double, NDIMS>& bb,
                                         const conduit::Node& node)
 {
-  int dim = node.fetch_existing("dim").as_int();
-  SLIC_ASSERT(dim == NDIMS);
-  const double* lo = node.fetch_existing("lo").as_double_ptr();
-  const double* hi = node.fetch_existing("hi").as_double_ptr();
-  bb =
```
Comment (kennyweiss) [inline src/axom/quest/DistributedClosestPoint.hpp:148]: Fun fact -- loading it this way was actually setting the "uninitialized" bounding box to have infinite extent ("lo" was the smallest double; "hi" was the largest double), instead of invalid bounds!  My bugfix resolves this.

File: src/axom/quest/DistributedClosestPoint.hpp [hunk 9]
```diff
@@ -686,9 +716,12 @@ class DistributedClosestPointImpl
         int nextDst = (m_rank + 1) % m_nranks;
         get_bounding_box_from_conduit_node(queryPartitionBb,
                                            xferNodePtr->fetch_existing("aabb"));
+
         while(xferNodePtr && nextDst != m_rank)
         {
-          double sqDist =
+          const bool isValidDistance = queryPartitionBb.isValid() &&
+            m_objectPartitionBbs[nextDst].isValid();
```
Comment (kennyweiss) [inline src/axom/quest/DistributedClosestPoint.hpp:723]: Since this explicitly checks for invalid bounding boxes, it is independent of the decision regarding the return value of `squared_distance(bbox,bbox)` for invalid bounding boxes (i.e. `nan` vs. max double).

File: src/axom/quest/DistributedClosestPoint.hpp (hunk 8 above)
Comment (gunney1) [reply src/axom/quest/DistributedClosestPoint.hpp:148]: Thanks @kennyweiss

File: src/axom/primal/operators/squared_distance.hpp (hunk 7 above)
Comment (gunney1) [reply src/axom/primal/operators/squared_distance.hpp:134]: Sounds good.  I'd like to go with whatever works best for the majority of users.

File: src/axom/primal/operators/squared_distance.hpp (hunk 6 above)
Comment (gunney1) [reply src/axom/primal/operators/squared_distance.hpp:83]: Ok

File: src/axom/quest/DistributedClosestPoint.hpp (hunk 9 above)
Comment (gunney1) [reply src/axom/quest/DistributedClosestPoint.hpp:723]: That makes it clear and robust.  Thanks @kennyweiss
